requests-for-comments/2024-05-28-monorepo.md

```diff
@@ -0,0 +1,46 @@
+# Request for comments: Monorepo - combine rust services and main posthog app
+
+## Problem statement
+*Who are we building for, what are their needs (include example cases), why is this important?*
+
+I've had a few conversations with everyone involved, and there hasn't been much pushback, so I'll very quickly summarise the problem, and then focus this RFC on how we get to what we want.
+
+Having things split up between hog-rs and posthog app make a couple things pretty problematic:
+1. Trying to develop locally with both, the capture-rs/decide-rs endpoints running locally, and the main posthog app running locally
+2. With shared data models (example: django app controlling flag CRUD and redis population, decide-rs using this data to evaluate flags) means it becomes much harder to keep things in sync, and making sure a data model update doesn't break the other service.
+3. It's harder to keep track of changes that need to be made everywhere / keeping code in sync across repos.
+4. We need to re-create similar scaffolding to make things work in the rust repo (example: GeoIP db downloading, webhooks data model setup, etc.)
+
+Now, I'm looking for alignment on the steps below, and whether this makes sense, and if I'm missing anything.
+
+## Design 
+*What are the key user experience and technical design decisions / trade-offs?*
+
+To make this work, here's what we need to do: (non-exhaustive, hoping you can add what I'm missing!)
+
+Effectively, nothing should change for anyone developing on `hog-rs` currently, and people developing on main `posthog` repo should get more options to hook things up.
+
+1. [ ] Introduce a `rust/` top-level folder that has all rust services
+2. [ ] Setup local dev options to either:
+    3. [ ] Use docker image for capture-rs/decide-rs, hooked up to local posthog app (when you don't care about capture-rs)
+    4. [ ] Use locally running capture-rs/decide-rs, hooked up to local posthog app (when you are developing e2e)
+    5. [ ] Use locally running capture-rs/decide-rs, independent of local posthog app (when focusing just on capture-rs/decide-rs)
+3. [ ] Modify pre-commit hook to handle `rust/` 
+
+### CI
+
+1. [ ] Keep hog-rs services fast - make sure random django tests don't run on CI for rust services, and vice-versa
+2. [ ] Copy over hog-rs deployment / docker-image build, and make sure this doesn't run on random django changes
+3. [ ] Figure out all main repo CI workflows that we need to now skip for the `rust/` folder, if any
+5. [ ] [Can be done a bit later] - setup e2e tests that run across rust services and main app to make sure things work properly together
+
+
+### Migration plan
+
+I'm not 100% sure yet how we want to migrate things over, or over what timeline.
+
+Given hog-rs deploys happen manually right now, it seems de-risked enough to do it in one fell-swoop to minimise keeping-code-in-two-places-in-sync-pains. So I'd propose:
```

Comment:
All sounds good, except this. TIL we have to deploy hog-rs manually. If this was embedded within `posthog/posthog` then I would have assumed that deployments were automatic with the rest of the codebase. With `hog-rs` being a separate repo, I would have intentionally seeked out how it got deployed

Reply:
Correct - eventually this will be automated, but for now it's good enough for hog-rs editors 👍